# Decoding dynamic cell-cell communication-driven cell state transitions in spatial transcriptomics

## Introduction
In multicellular systems, cell fate determination emerges from the integration of intracellular signaling and intercellular communication. Spatial transcriptomics (ST) provides unprecedented opportunities to elucidate these regulatory processes.

Here, we introduce CCCvelo, a computational framework designed to reconstruct CCC-driven CST dynamics by jointly optimizing a dynamic CCC signaling network and a latent CST clock. To achieve this, CCCvelo formulates a unified multiscale nonlinear kinetic model that integrates intercellular ligand-receptor signaling gradients with intracellular transcription factor activation cascades to capture gene expression dynamics encoding CSTs. Moreover, we devise PINN-CELL, a physics-informed neural network-based coevolution learning algorithm, which simultaneously optimizes model parameters and pseudotemporal ordering. Benchmarking and application of CCCvelo to synthetic data and high-resolution ST datasets, including mouse cortex, embryonic trunk development, and human prostate cancer datasets, demonstrate its ability to successfully recover known morphogenetic trajectories while uncovering dynamic CCC signaling rewiring that orchestrates CST progression. By integrating mechanistic CCC modeling with trajectory learning in spatially resolved omics data, CCCvelo provides a powerful framework for decoding the multiscale and dynamic regulatory principles that govern cell fate decisions.
